clients/src/main/java/org/apache/kafka/clients/consumer/internals/MemberAssignmentReconciler.java

```diff
@@ -0,0 +1,291 @@
+ * steps may result in one of the following:
+ *
+ * <ul>
+ *     <li>
+ *         {@link ReconciliationResult#NO_CHANGE}: no changes were made to the set of partitions.
+ *     </li>
+ *     <li>
+ *         {@link ReconciliationResult#IN_PROGRESS}: changes to the assignment have started. In practice this means
+ *         that the appropriate {@link ConsumerRebalanceListener} callback method is being invoked.
+ *     </li>
+ *     <li>
+ *         {@link ReconciliationResult#COMPLETED}: the {@link ConsumerRebalanceListener} callback method was made and
+ *         the changes were applied locally.
+ *     </li>
+ *     <li>
+ *         {@link ReconciliationResult#EXPIRED}: something happened to cause the operation to modify the assigned set
+ *         of partitions. This could be caused by a {@link ConsumerRebalanceListener} callback method that takes too
+ *         long to execute, interruption with the consumer group coordinator, or other reasons.
+ *     </li>
+ * </ul>
+ *
+ * The comparison against the {@link SubscriptionState#assignedPartitions() current set of assigned partitions} and
+ * the {@link Assignment#assignedTopicPartitions() target set of assigned partitions} is performed by essentially
+ * <em>flattening</em> the respective entries into two sets of {@link org.apache.kafka.common.TopicPartition partitons}
+ * which are then compared using basic {@link Set} comparisons.
+ */
+public class MemberAssignmentReconciler {
+
+    /**
+     * The result of the {@link #revoke(Optional, Timer)} or {@link #assign(Optional, Timer)} methods being invoked.
+     */
+    enum ReconciliationResult {
+        NO_CHANGE,
+        IN_PROGRESS,
+        COMPLETED,
+        EXPIRED
+    }
+
+    // Ugly little handler enum for making logging less verbose.
+    private enum Operation {
+
+        REVOKE("revoke", "revoked"), ASSIGN("assign", "assigned");
+
+        private final String verbPastTense;
+
+        private final String methodName;
+
+        Operation(String verbPresentTense, String verbPastTense) {
+            this.verbPastTense = verbPastTense;
+            this.methodName = String.format("%s.onPartitions%s()", ConsumerRebalanceListener.class.getSimpleName(), verbPresentTense.substring(0, 1).toUpperCase(Locale.ROOT) + verbPresentTense.substring(1).toLowerCase(Locale.ROOT));
+        }
+    }
+
+    private final Logger log;
+    private final SubscriptionState subscriptions;
+    private final ConsumerMetadata metadata;
+    private final BlockingQueue<BackgroundEvent> backgroundEventQueue;
+    private Optional<RebalanceCallbackEvent> inflightCallback;
+
+    public MemberAssignmentReconciler(LogContext logContext,
+                                      SubscriptionState subscriptions,
+                                      ConsumerMetadata metadata,
+                                      BlockingQueue<BackgroundEvent> backgroundEventQueue) {
+        this.log = logContext.logger(getClass());
```

Comment:
I wonder what's the convention for the use of `this`.  I got pointed out several times by jun for not using it.

Reply:
Use of `this` when assigning instance variables is pretty common. Use of `this` to refer to the current object under other situations isn't strictly essential.